doc: Add configuration explanation (SC-1169) (#1749)
canonical/cloud-init: doc/rtd/topics/configuration.rst

.. _configuration:

Configuration
*************

Internally, cloud-init builds a single configuration that is then referenced
throughout the life of cloud-init. The configuration is built from multiple
sources such that if a key is defined in multiple sources, the higher priority
source overwrites the lower priority source.

From lowest priority to highest, configuration sources are:

* **Hardcoded config**: Config_ that lives within the source of cloud-init
  and cannot be changed.
* **Configuration directory**: Anything defined in ``/etc/cloud/cloud.cfg`` and

>   ``/etc/cloud/cloud.cfg.d``

Reviewer: ```suggestion
  ``/etc/cloud/cloud.cfg.d``.
```

> * **Runtime config**: Anything defined in ``/run/cloud-init/cloud.cfg``

Reviewer: ```suggestion
* **Runtime config**: Anything defined in ``/run/cloud-init/cloud.cfg``.
```

* **Kernel command line**: On the kernel command line, anything found between
  ``cc:`` and ``end_cc`` will be interpreted as cloud-config user data.

These four sources make up the **base configuration**. Added to these
to provide the full configuration are:

* **Vendor data**: :ref:`Data<vendordata>` provided by the datasource
* **User data**: :ref:`Data<user_data_formats>` also provided by
  the datasource

> .. note::

Reviewer: I found this note a bit difficult to understand. 

As said in the previous paragraph, base configuration is merged from 4 sources, not only `/etc/cloud/cloud.cfg`. Reading this, I ask my self, have the other 3 sources to follow _the specific user data format_?

Maybe it's worth rephrasing this to something in the lines of:

base-configuration sources do not conform to :ref:`#cloud-config<topics/format:Cloud Config Data>`, ...

Author: Thanks. I didn't love the wording either, so you make a good point. The impetus for it is that before we had jinja templating in /etc/cloud*, multiple people have confused the #cloud-config mentioned [here](https://cloudinit.readthedocs.io/en/latest/topics/instancedata.html#using-instance-data) with /etc/cloud/cloud.cfg. Part of that confusion probably has something to do with them both being casually referred to as "cloud config".

I update that section as well.

  Even though baseconfiguration is found in ``/etc/cloud/cloud.cfg``, it is
  distinct from :ref:`#cloud-config<topics/format:Cloud Config Data>`, which
  is a specific user data format.

Network Configuration
=====================
Network configuration happens independently from other cloud-init

> configuration. It is sourced from the cloud provided metadata servce and is

Reviewer: ```suggestion
configuration. It is sourced from the cloud provided metadata service and is
```

.. _configuration:

Configuration Sources
*********************

Internally, cloud-init builds a single configuration that is then referenced
throughout the life of cloud-init. The configuration is built from multiple
sources such that if a key is defined in multiple sources, the higher priority
source overwrites the lower priority source.

Base Configuration
==================

From lowest priority to highest, configuration sources are:

* **Hardcoded config**: Config_ that lives within the source of cloud-init
  and cannot be changed.
* **Configuration directory**: Anything defined in ``/etc/cloud/cloud.cfg`` and
  ``/etc/cloud/cloud.cfg.d``.
* **Runtime config**: Anything defined in ``/run/cloud-init/cloud.cfg``.
* **Kernel command line**: On the kernel command line, anything found between
  ``cc:`` and ``end_cc`` will be interpreted as cloud-config user data.

These four sources make up the base configuration.

Vendor and User Data
====================
Added to the base configuration are:

> * **Vendor data**: :ref:`Data<vendordata>` provided by the datasource

Reviewer 2: Should we also reference metadata here too?

Author: Is it? What metadata do we fetch from clouds that can be overridden by vendor data?

Reviewer 2: I was thinking specifically of `local_hostname` in metadata is the most likely case where metadata gets overridden by a #cloud-config `hostname` value in either vendor-data or user-data. I can't think of more explicit examples off the top of my head but DataSource does provide a get_config_obj which some datasources (Azure) provide which returns ds.cfg and will participate in the merged config via [ConfigMerger._read_cfg](https://github.com/canonical/cloud-init/blob/main/cloudinit/helpers.py#L263-L282). I think generally it's not frequently a case where that ds.cfg data participates, but there is a mechanism that sets priority at datasource metadata config, vendor data then user data. Not worth holding up this PR as-is, but something we can iterate on more in the future if needed.

* **User data**: :ref:`Data<user_data_formats>` also provided by
  the datasource

These get fetched from the datasource and are defined at instance launch.

.. note::
  While much of what is defined in the base configuration can be overridden by
  vendor data and user data, base configuration sources do not conform to
  :ref:`#cloud-config<topics/format:Cloud Config Data>`

Network Configuration
=====================
Network configuration happens independently from other cloud-init
configuration. It is sourced from the cloud provided metadata service and is
not user editable. If you need to change your network configuration,
your options are:

* Use the cloud-provided tools to update your network topology in the cloud.

> * :ref:`Disable cloud-init network configuration entirely

Reviewer 2: A third option here is to provide network configuration in a customized image directly in `/etc/cloud/cloud.cfg.d/*.cfg` with a network: <<config>> definition.

  <topics/network-config:Disabling Network Configuration>`.
  This leaves network configuration entirely up to you.

Specifying Configuration
==========================

End users
---------
Pass :ref:`user data<user_data_formats>` to the cloud provider.
Every platform supporting cloud-init will provide a method of supplying
user data. If you're unsure how to do this, reference the documentation
provided by the cloud platform you're on. Additionally, there may be
related cloud-init documentation in the :ref:`datasource<datasources>`
section.

Once an instance has been initialized, the user data may not be edited.
It is sourced directly from the cloud, so even if you find a local file
that contains user data, it will likely be overwritten next boot.

> To modify the run frequency of a particular cloud-init module,
> :ref:`cloud-init single<cli_single>` may be used. This can be used to
> disable a module you no longer with to run.

Reviewer 2: This is not true is it?  The CLI doesn't modify the run frequency of a module, it allows you to override that frequency manually on individual calls.

Maybe something like:
```suggestion
To override the default run frequency of a particular cloud-init module,
:ref:`cloud-init single<cli_single>` may be invoked from scripts, cloud-config `runcmd` or `boodcmd` sections .
 ```

Not sure we want to document the following but:

To modify the default frequency of a particular cloud-init module to `once-per-instance`, `always`, or `once`. Base config in /etc/cloud/cloud.cfg(.d) is needed. To override default frequency for a particular module, provide a list of values for the specific module name of the format `[<module_name>, <frequency>]` in any of the cloud-init boot stages `cloud_init_modules`, `cloud_config_modules` or `cloud_final_modules`:

```
cloud_init_modules:
 - [ write-files, once-per-instance]
....
```

Author: Doh, good catch. I misread the documentation for cloud-init single, so I've also updated that documentation to be slightly clearer. I also removed this bit for reasons I'll explain in a larger comment on the PR.

Distro Providers
----------------
Modify the base config. This often involves submitting a PR to modify
the base `cloud.cfg template`_, which is used to customize
`/etc/cloud/cloud.cfg` per distro. Additionally, a file can be added to
``/etc/cloud/cloud.cfg.d`` to override a piece of the base configuration.

Cloud Providers
---------------

> Pass vendor data. This is the preferred method for clouds to provide

Reviewer 2: worth also mentioning meta-data provided by clouds here?

Author: Same comment as above. I'm not sure how other metadata is relevant here.
